File: src/doctor.ts
import { access, constants, mkdir, rm, writeFile } from "node:fs/promises";
import { dirname } from "node:path";
import type { Config } from "./config.ts";
import { needsMaster } from "./config.ts";
import { ensureDumpsRootWritable, resolveDumpsRoot } from "./dumps.ts";
import { assertDockerAvailable } from "./docker.ts";
import { dbPathForConfig, vaultHasMaster } from "./metadata.ts";
import { getEncId, openVault } from "./vault.ts";

export type DoctorCheck = {
  name: string;
  ok: boolean;
  message: string;
  hint?: string;
};

export type DoctorReport = {
  ok: boolean;
  checks: DoctorCheck[];
};

async function dockerVersion(): Promise<string> {
  try {
    const proc = Bun.spawn(["docker", "--version"], {
      stdout: "pipe",
      stderr: "pipe",
      stdin: "ignore",
    });
    const [stdout, exitCode] = await Promise.all([
      new Response(proc.stdout).text(),
      proc.exited,
    ]);
    if (exitCode !== 0) return "docker present (version unknown)";
    return stdout.trim() || "docker present";
  } catch {
    return "docker present (version unknown)";
  }
}

async function probeParentDir(parent: string): Promise<void> {
  await mkdir(parent, { recursive: true });
  try {
    await access(parent, constants.W_OK);
  } catch {
    throw new Error(`Parent directory is not writable: ${parent}`);
  }
  const probe = `${parent}/.dumpmgr-doctor-${Date.now()}`;
  try {
    await writeFile(probe, "ok");
  } finally {
    await rm(probe, { force: true });
  }
}

async function checkVaultDb(vaultPath: string): Promise<{
  ok: boolean;
  message: string;
  hint?: string;
}> {
  if (!(await Bun.file(vaultPath).exists())) {
    return {
      ok: false,
      message: `vault database not found at ${vaultPath}`,
      hint: "run `dumpmgr config init`",
    };
  }
  return { ok: true, message: `vault database OK (${vaultPath})` };
}

export async function runDoctor(
  config: Config,
  configPath: string,
): Promise<DoctorReport> {
  const checks: DoctorCheck[] = [];

  try {
    await assertDockerAvailable();
    const ver = await dockerVersion();
    checks.push({ name: "docker", ok: true, message: ver });
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    checks.push({
      name: "docker",
      ok: false,
      message,
      hint: "start Docker Desktop / the docker daemon and retry",
    });
  }

  const dumpsRoot = resolveDumpsRoot(config.dumpDirectory);
  const parent = dirname(dumpsRoot);
  try {
    await probeParentDir(parent);
    checks.push({
      name: "dumps-parent",
      ok: true,
      message: `parent dir writable: ${parent}`,
    });
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    checks.push({
      name: "dumps-parent",
      ok: false,
      message,
    });
  }

  try {
    await ensureDumpsRootWritable(dumpsRoot);
    checks.push({
      name: "dumps-root",
      ok: true,
      message: `dumps dir writable: ${dumpsRoot}`,
    });
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    checks.push({ name: "dumps-root", ok: false, message });
  }

  const vaultPath = dbPathForConfig(configPath);
  const vaultCheck = await checkVaultDb(vaultPath);
  checks.push({ name: "vault-db", ...vaultCheck });

  if (config.s3Options) {
    checks.push({
      name: "s3-config",
      ok: true,
      message: `S3 configured: ${config.s3Options.endpoint}/${config.s3Options.bucketName}`,
      hint: "S3 credentials are checked by `dumpmgr s3 upload` or `dumpmgr s3 download` after master unlock.",
    });
  }

  if (vaultCheck.ok) {
    try {
      const db = await openVault(configPath);
      try {
        if (needsMaster(config)) {
          const has = await vaultHasMaster(configPath);
          checks.push({
            name: "kdf-salt",
            ok: has,
            message: has ? "kdfSalt present" : "kdfSalt missing",
            hint: has ? undefined : "run `dumpmgr config init` to set a master password",
          });
          checks.push({
            name: "master-hash",
            ok: has,
            message: has ? "master password hash present" : "master password hash missing",
            hint: has ? undefined : "run `dumpmgr config init` to set a master password",
          });
        }
        if (config.encryptedDump) {
          const id = getEncId(db);
          checks.push({
            name: "enc-id",
            ok: Boolean(id),
            message: id ? `encId present (${id})` : "encId missing",
            hint: id ? undefined : "encId is generated on next successful master unlock",
          });
        }
      } finally {
        db.close();
      }
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      checks.push({
        name: "vault-body",
        ok: false,
        message: `cannot open vault: ${message}`,
      });
    }
  }

  const ok = checks.every((c) => c.ok);
  return { ok, checks };
}

File: src/internal/doctor/doctor.go
package doctor

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/md-redwan-hossain/dumpmgr/src/internal/config"
	"github.com/md-redwan-hossain/dumpmgr/src/internal/docker"
	"github.com/md-redwan-hossain/dumpmgr/src/internal/dumps"
	"github.com/md-redwan-hossain/dumpmgr/src/internal/metadata"
	"github.com/md-redwan-hossain/dumpmgr/src/internal/vault"
)

type Check struct {
	Name    string
	OK      bool
	Message string
	Hint    string
}

type Report struct {
	OK     bool
	Checks []Check
}

func Run(cfg *config.Config, configPath string) Report {
	var checks []Check

	if err := docker.AssertAvailable(); err != nil {
		checks = append(checks, Check{
			Name:    "docker",
			OK:      false,
			Message: err.Error(),
			Hint:    "start Docker Desktop / the docker daemon and retry",
		})
	} else {
		ver, _ := docker.Version()
		checks = append(checks, Check{Name: "docker", OK: true, Message: ver})
	}

	dumpsRoot := dumps.ResolveRoot(cfg.DumpDirectory)
	parent := filepath.Dir(dumpsRoot)
	if err := probeParentDir(parent); err != nil {
		checks = append(checks, Check{Name: "dumps-parent", OK: false, Message: err.Error()})
	} else {
		checks = append(checks, Check{
			Name:    "dumps-parent",
			OK:      true,
			Message: fmt.Sprintf("parent dir writable: %s", parent),
		})
	}

	if err := dumps.EnsureRootWritable(dumpsRoot); err != nil {
		checks = append(checks, Check{Name: "dumps-root", OK: false, Message: err.Error()})
	} else {
		checks = append(checks, Check{
			Name:    "dumps-root",
			OK:      true,
			Message: fmt.Sprintf("dumps dir writable: %s", dumpsRoot),
		})
	}

	vaultPath := metadata.DBPathForConfig(configPath)
	vaultCheck := checkVault(vaultPath)
	checks = append(checks, Check{Name: "vault-db", OK: vaultCheck.OK, Message: vaultCheck.Message, Hint: vaultCheck.Hint})

	if cfg.S3Options != nil {
		checks = append(checks, Check{
			Name:    "s3-config",
			OK:      true,
			Message: fmt.Sprintf("S3 configured: %s/%s", cfg.S3Options.Endpoint, cfg.S3Options.BucketName),
			Hint:    "S3 credentials are checked by `dumpmgr s3 upload` or `dumpmgr s3 download` after master unlock.",
		})
	}

	if vaultCheck.OK {
		store, err := vault.Open(configPath)
		if err != nil {
			checks = append(checks, Check{
				Name:    "vault-body",
				OK:      false,
				Message: fmt.Sprintf("cannot open vault: %v", err),
			})
		} else {
			defer store.Close()
			if config.NeedsMaster(cfg) {
				hasMaster, _ := store.HasMaster()
				checks = append(checks, masterHashCheck(hasMaster))
				checks = append(checks, saltCheck(hasMaster))
			}
			if cfg.EncryptedDump {
				encID, _ := store.EncID()
				if encID != "" {
					checks = append(checks, Check{
						Name:    "enc-id",
						OK:      true,
						Message: fmt.Sprintf("encId present (%s)", encID),
					})
				} else {
					checks = append(checks, Check{
						Name:    "enc-id",
						OK:      false,
						Message: "encId missing",
						Hint:    "encId is generated on next successful master unlock",
					})
				}
			}
		}
	}

	ok := true
	for _, c := range checks {
		if !c.OK {
			ok = false
			break
		}
	}
	return Report{OK: ok, Checks: checks}
}

func probeParentDir(parent string) error {
	if err := os.MkdirAll(parent, 0o755); err != nil {
		return err
	}
	probe := filepath.Join(parent, fmt.Sprintf(".dumpmgr-doctor-%d", os.Getpid()))
	if err := os.WriteFile(probe, []byte("ok"), 0o644); err != nil {
		return fmt.Errorf("parent directory is not writable: %s", parent)
	}
	_ = os.Remove(probe)
	return nil
}

type vaultResult struct {
	OK      bool
	Message string
	Hint    string
}

func checkVault(vaultPath string) vaultResult {
	if _, err := os.Stat(vaultPath); os.IsNotExist(err) {
		return vaultResult{
			OK:      false,
			Message: fmt.Sprintf("vault database not found at %s", vaultPath),
			Hint:    "run `dumpmgr config init`",
		}
	} else if err != nil {
		return vaultResult{OK: false, Message: fmt.Sprintf("cannot read vault at %s", vaultPath)}
	}
	return vaultResult{OK: true, Message: fmt.Sprintf("vault database OK (%s)", vaultPath)}
}

func saltCheck(ok bool) Check {
	if ok {
		return Check{Name: "kdf-salt", OK: true, Message: "kdfSalt present"}
	}
	return Check{
		Name:    "kdf-salt",
		OK:      false,
		Message: "kdfSalt missing",
		Hint:    "run `dumpmgr config init` to set a master password",
	}
}

func masterHashCheck(ok bool) Check {
	if ok {
		return Check{Name: "master-hash", OK: true, Message: "master password hash present"}
	}
	return Check{
		Name:    "master-hash",
		OK:      false,
		Message: "master password hash missing",
		Hint:    "run `dumpmgr config init` to set a master password",
	}
}
